### PR TITLE
Add option to normalize loss per target

### PR DESCRIPTION
**Pull request by Muennighoff**

### CODE REVIEW

File: megatron/model/gpt_model.py [hunk 1]
```diff
@@ -190,7 +201,11 @@ def CrossEntropy(output, labels):
             expected_number_of_tokens = loss_mask.sum()
 
         loss_mask = loss_mask.view(-1)
-        loss = torch.sum(losses.view(-1) * loss_mask) / expected_number_of_tokens
+
+        loss_mask = fast_normalize(loss_mask)
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:205]: Note this is a breaking change, if you want to merge on the main repo you need to expose this via a CLI flag.

File: megatron/model/gpt_model.py [hunk 2]
```diff
@@ -190,7 +201,11 @@ def CrossEntropy(output, labels):
...
+        expected_num_of_target_seqs = loss_mask.sum()
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:206]: You compute this twice

File: megatron/arguments.py [hunk 3]
```diff
@@ -926,6 +926,8 @@ def __call__(self, parser, args, values, option_string=None):
                        help='Mask loss for the end of document tokens.')
     group.add_argument('--loss-on-targets-only', action='store_true',
                        help='Mask loss on input sequence.')
+    group.add_argument('--norm-target-loss', action='store_true',
+                       help='Normalize the loss per target.')
```
Comment (thomasw21) [inline megatron/arguments.py:930]: Can you improve the `help`? Meg-DS is used for pretraining and the notion of target doesn't exist and can be confused with labels which is very different. Maybe one way would be to create `finetuning` options in another group where we clearly define terminology.

File: megatron/model/gpt_model.py [hunk 4]
```diff
@@ -159,7 +158,18 @@ def load_state_dict(self, state_dict, strict=True):
         self.language_model.load_state_dict(state_dict, strict=strict)
 
 
-def get_cross_entropy(is_prefix: bool):
+def fast_normalize(loss_mask: torch.Tensor):
+    """
+    Turn loss_mask from [0,0,0,1,1,0,0,1,0,0,1,1,1] > [0,0,0,0.5,0.5,0,0,1,0,0,0.3,0.3,0.3]
+
+    Credits to @thomasw21 for this efficient implementation!
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:165]:
```suggestion
```

Haha thanks for the mention but no need to leave in the codebase.

File: megatron/model/gpt_model.py [hunk 5]
```diff
@@ -159,7 +158,18 @@ def load_state_dict(self, state_dict, strict=True):
...
+    """
+    loss_mask = loss_mask.float()
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:167]: Hum is this needed? what's `loss_mask` dtype usually? bool?

File: megatron/model/gpt_model.py [hunk 6]
```diff
@@ -186,10 +196,15 @@ def CrossEntropy(output, labels):
             else:
                 average_tokens_per_sample = sequence_length
             expected_number_of_tokens = average_tokens_per_sample * micro_batch_size
+        elif norm_target_loss:
+            loss_mask = loss_mask.view(-1)
+            loss_mask = fast_normalize(loss_mask)
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:201]: Why not build this in `get_batch_pipe`?

File: megatron/model/gpt_model.py (hunk 5 above)
Comment (Muennighoff) [reply megatron/model/gpt_model.py:165]: Yeah it's bool

File: megatron/model/gpt_model.py (hunk 6 above)
Comment (Muennighoff) [reply megatron/model/gpt_model.py:199]: Is there any disadvantage to doing it here?
I think it's easier to understand if it's in the sample place.

Comment (thomasw21) [reply megatron/model/gpt_model.py:199]: I think it's weird to construct half of the `loss_mask` somewhere, and another somewhere else.

File: megatron/model/gpt_model.py (hunk 5 above)
Comment (Muennighoff) [reply megatron/model/gpt_model.py]: Sorry you're right; removed the cast

File: finetune_t0.py [hunk 7]
```diff
@@ -95,6 +103,10 @@ def get_batch_pipe(data):
         segment_ids=segment_ids.long(),
     )
 
+    if args.norm_target_loss:
+        loss_mask = loss_mask.view(-1)
```
Comment (thomasw21) [inline finetune_t0.py:107]: There's a fun hack you can do, `view` have the same storage space as the initial model. so you can probably write something like:

```
def fast_normalize(loss_mask: torch.Tensor):
    """
    Turn loss_mask from [0,0,0,1,1,0,0,1,0,0,1,1,1] > [0,0,0,0.5,0.5,0,0,1,0,0,0.3,0.3,0.3]
    """
    flatten_view = loss_mask.view(-1)
    _, inverse_indices, counts = torch.unique_consecutive(loss_mask, return_inverse=True, return_counts=True)
    counts = torch.gather(dim=0, index=inverse_indices, input=counts)
    flatten_view.div_(counts)
    return loss_mask
```

Comment (thomasw21) [reply finetune_t0.py:107]: you could also `clone` before doing this operation so that you actually don't make `fast_normalize` a in-place operation.

File: megatron/model/gpt_model.py [hunk 8]
```diff
@@ -186,10 +185,13 @@ def CrossEntropy(output, labels):
...
                 average_tokens_per_sample = sequence_length
             expected_number_of_tokens = average_tokens_per_sample * micro_batch_size
+        elif args.norm_target_loss:
+            expected_num_of_target_seqs = loss_mask.sum()
+            loss = torch.sum(losses.view(-1) * loss_mask) / expected_num_of_target_seqs
+            return loss
         else:
             expected_number_of_tokens = loss_mask.sum()
 
-        loss_mask = loss_mask.view(-1)
```
Comment (thomasw21) [inline megatron/model/gpt_model.py:192]: I'm pretty sure that breaks everything. Can you run tests?

File: finetune_t0.py (hunk 7 above)
Comment (Muennighoff) [reply finetune_t0.py:107]: Why is

```
def fast_normalize(loss_mask: torch.Tensor):
    """
    Turn loss_mask from [0,0,0,1,1,0,0,1,0,0,1,1,1] > [0,0,0,0.5,0.5,0,0,1,0,0,0.3,0.3,0.3]
    """
    flatten_view = loss_mask.view(-1)
    _, inverse_indices, counts = torch.unique_consecutive(loss_mask, return_inverse=True, return_counts=True)
    counts = torch.gather(dim=0, index=inverse_indices, input=counts)
    flatten_view.div_(counts)
    return loss_mask
```

better than
```
def fast_normalize(loss_mask: torch.Tensor):
    """
    Turn loss_mask from [0,0,0,1,1,0,0,1,0,0,1,1,1] > [0,0,0,0.5,0.5,0,0,1,0,0,0.3,0.3,0.3]
    """
    _, inverse_indices, counts = torch.unique_consecutive(loss_mask, return_inverse=True, return_counts=True)
    counts = torch.gather(dim=0, index=inverse_indices, input=counts)
    return loss_mask / counts
```
?

Comment (thomasw21) [reply finetune_t0.py:107]: Does the latter work if `loss_mask` is not 1D?
